### ISSUE TITLE
Android 8.0 crash

### ISSUE THREAD

**Issue by drewster99** [OneSignal/OneSignal-Android-SDK#337]
Seeing these on Android 8 devices, such as Pixel:

`oday, 9:37 AM on app version 313
Google Pixel (sailfish), 4096MB RAM, Android 8.0
Report 1

java.lang.RuntimeException: 
  at android.app.ActivityThread.handleReceiver (ActivityThread.java:3259)
  at android.app.ActivityThread.-wrap17 (Unknown Source)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1677)
  at android.os.Handler.dispatchMessage (Handler.java:105)
  at android.os.Looper.loop (Looper.java:164)
  at android.app.ActivityThread.main (ActivityThread.java:6541)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.Zygote$MethodAndArgsCaller.run (Zygote.java:240)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:767)

Caused by: java.lang.IllegalStateException: 
  at android.app.ContextImpl.startServiceCommon (ContextImpl.java:1505)
  at android.app.ContextImpl.startService (ContextImpl.java:1461)
  at android.content.ContextWrapper.startService (ContextWrapper.java:644)
  at android.content.ContextWrapper.startService (ContextWrapper.java:644)
  at android.support.v4.content.WakefulBroadcastReceiver.startWakefulService (WakefulBroadcastReceiver.java:89)
  at **com.onesignal.GcmBroadcastReceiver.onReceive (GcmBroadcastReceiver.java:64)**
  at android.app.ActivityThread.handleReceiver (ActivityThread.java:3252)`

**Comment by Alamgir** [comment]
@drewster99 thanks for letting us know about this issue - we're currently looking at upgrading the SDK at various points to be O-compatible, including the move away from WakefulBroadcastReceiver

**Comment by Alamgir** [comment]
@drewster99 what version of the SDK are you using?

**Comment by drewster99** [comment]
Hmm.. good question.. is there an easy way to check?

Sent from my iPhone

On Oct 19, 2017, at 4:18 PM, Alamgir Mand <notifications@github.com> wrote:

@drewster99 what version of the SDK are you using?

**Comment by jkasten2** [comment]
@drewster99 You can check the version you have in you `build.gradle`. You can also check on the All Users page on the OneSignal dashboard to check the exact version in production. You will have to enable "SDK Version" from the "VISIBLE COLUMN LIST" to see it though.

**Comment by drewster99** [comment]
I've got a Unity plug-in version of some sort.  The build.gradle isn't helpful as I'm just including onesignal-unity.aar.

**Comment by drewster99** [comment]
Ahh thanks for that!  -- The dashboard shows "020602"..  I assume that's 2.6.2?  I assume you're going to recommend I try updating.  I'm using the Unity plug-in -- what's the best way of updating it without completely trashing my config?

Also, I know I saw a few other issues here where people had mentioned Oreo/Android 8.  Is the latest version the best to update to?  Or is there some older "stable" build?

Thx!

**Comment by jkasten2** [comment]
@drewster99 If you are still targeting Android 7. 1 (API Level 25) or older then this shouldn't be an issue.

Unless the user opens Settings > Apps > App Info. The selects Battery > "Background activity" and disable this. I believe this would be done rarely by users but you might see a few crash reports from it. More details below:
http://www.androidpolice.com/2017/08/22/android-oreo-feature-spotlight-can-force-androids-new-background-limits-old-apps/

Going to the Unity 1.X version you are on to the latest [2.4.1](https://github.com/OneSignal/OneSignal-Unity-SDK/releases) version will take some changes however it is recommend due to the number of other fixes this contains as well.
I recommend removing the old OneSignal SDK files from your project before updating. Leave the OneSignal calls themselves in your code. Then follow the new setup guide below updating your init call to OneSignal.
https://documentation.onesignal.com/docs/unity-sdk-setup

**Comment by n1te1337** [comment]
@jkasten2 I'm targeting `22` and I'm running `3.6.5` and the app crashes on my pixel with the same error.

Happy to provide additional info if needed

**Comment by jkasten2** [comment]
@n1te1337 Can you provide a full stack trace from the logcat from the crash? The same flow should not be happening on the 3.5.6 SDK.

**Comment by n1te1337** [comment]
@jkasten2 on a second thought the stack track is slightly different:
```
Google Pixel (sailfish), 4096MB RAM, Android 8.0
Report 1

java.lang.RuntimeException: 
  at android.app.ActivityThread.handleReceiver (ActivityThread.java:3259)
  at android.app.ActivityThread.-wrap17 (Unknown Source)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1677)
  at android.os.Handler.dispatchMessage (Handler.java:105)
  at android.os.Looper.loop (Looper.java:164)
  at android.app.ActivityThread.main (ActivityThread.java:6541)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.Zygote$MethodAndArgsCaller.run (Zygote.java:240)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:767)
Caused by: java.lang.IllegalArgumentException: 
  at android.os.Parcel.readException (Parcel.java:1946)
  at android.os.Parcel.readException (Parcel.java:1888)
  at android.app.job.IJobScheduler$Stub$Proxy.enqueue (IJobScheduler.java:211)
  at android.app.JobSchedulerImpl.enqueue (JobSchedulerImpl.java:53)
  at com.onesignal.JobIntentService$JobWorkEnqueuer.enqueueWork (JobIntentService.java:337)
  at com.onesignal.JobIntentService.enqueueWork (JobIntentService.java:515)
  at com.onesignal.NotificationBundleProcessor.startExtenderService (NotificationBundleProcessor.java:434)
  at com.onesignal.NotificationBundleProcessor.processBundleFromReceiver (NotificationBundleProcessor.java:400)
  at com.onesignal.GcmBroadcastReceiver.processOrderBroadcast (GcmBroadcastReceiver.java:110)
  at com.onesignal.GcmBroadcastReceiver.onReceive (GcmBroadcastReceiver.java:67)
  at android.app.ActivityThread.handleReceiver (ActivityThread.java:3252)
```

**Comment by jkasten2** [comment]
@n1te1337 It could be related to issue #351. Can you try adding `android:permission="android.permission.BIND_JOB_SERVICE"` to your `NotificationExtender` in your `AndroidManifest.xml`?

**Comment by n1te1337** [comment]
@jkasten2 just added it. Is there a known way to trigger the error, or should it just crash on every notification?

**Comment by jkasten2** [comment]
@n1te1337 Not sure what the root cause of the issue is. It's crashing from a Java class built into Android it self `IJobScheduler.java` and isn't providing any specific details. If it was related to `android.permission.BIND_JOB_SERVICE` it should crash every time if you are missing it.

Since you are targeting `22` it is possible that the crash stack would be different then when targeting `26`.

Were you able to reproduce the crash on your test devices at any point?

**Comment by n1te1337** [comment]
Nope, wasn't able to reproduce it, it would just crash intermittently so I assumed that it was caused by something in my code until I saw this _issue_. I've uploaded the new version to the Pixel, in the past it would crash every 1-2 hours so I'll update here a bit later with the results.

Thanks for your help.

**Comment by n1te1337** [comment]
@jkasten2 no crashes in 24 hrs so it looks like adding `android.permission.BIND_JOB_SERVICE` fixes the issue. Thanks

**Comment by thariq12** [comment]
Hi, I'm using MI A1 smart phone with Oreo. I receive error notifications very often when the phone receive  a notification and the details section starts with "java.lang.runtimeexception" and some where in the middle it shows "com.onesignal.gcm", but I do not use or signed up to One signal services and I never tried. how can I remove the pop up?
![screenshot_20180512-124230](https://user-images.githubusercontent.com/31386353/39954770-290c34e2-55e2-11e8-8153-00aeb6503e09.png)

**Comment by jkasten2** [comment]
@thariq12 The developer of the app `weave.tech.libretorrent` you are using didn't seem to update their app to support Android 8 (Oreo) devices. You will need to send this error to the app creators so they can update their app.
